src/main/java/org/jmxtrans/agent/Query.java

```diff
@@ -131,20 +131,35 @@ public void collectAndExport(@Nonnull MBeanServer mbeanServer, @Nonnull OutputWr
 
         for (ObjectName on : objectNames) {
             try {
-                if (attribute == null || attribute.isEmpty()) {
-                    for (MBeanAttributeInfo mBeanAttributeInfo : mbeanServer.getMBeanInfo(on).getAttributes()) {
-                        collectAndExportAttribute(mbeanServer, outputWriter, on, mBeanAttributeInfo.getName());
-                    }
-
-                } else {
-                    collectAndExportAttribute(mbeanServer, outputWriter, on, attribute);
-                }
+                collectAndExportForObjectName(mbeanServer, on, outputWriter);
             } catch (Exception e) {
                 logger.log(Level.WARNING, "Exception collecting " + on + "#" + attribute + (key == null ? "" : "#" + key), e);
             }
         }
     }
 
+    private void collectAndExportForObjectName(MBeanServer mbeanServer, ObjectName on, OutputWriter outputWriter)
+            throws Exception {
+        for (String attribute : resolveAttributes(mbeanServer, on)) {
+                collectAndExportAttribute(mbeanServer, outputWriter, on, attribute);
+        }
+    }
+
+    private List<String> resolveAttributes(MBeanServer mbeanServer, ObjectName on) throws Exception {
+        List<String> resolvedAttributes = new ArrayList<>();
+        if (attribute == null || attribute.isEmpty()) {
+            // Null or empty attribute specified, collect all attributes
+            for (MBeanAttributeInfo mBeanAttributeInfo : mbeanServer.getMBeanInfo(on).getAttributes()) {
+                resolvedAttributes.add(mBeanAttributeInfo.getName());
+            }
+        } else {
+           // Single attribute or comma-separated list specified, collect specified attribute(s).
+           String[]  attributes = attribute.split(",");
```

Comment:
Could we just trim the value of the splitted `attributes` so that we don't have backward compatibility issues?

A bit like Google Guava's Splitter
https://code.google.com/p/guava-libraries/wiki/StringsExplained

> Splitter.on(',').trimResults().split("a, b, c, d") returns "a", "b", "c", "d"